core/sr-io/without_std.rs

```diff
@@ -475,17 +475,17 @@ pub fn storage_root() -> [u8; 32] {
 }
 
 /// "Commit" all existing operations and compute the resultant child storage root.
-pub fn child_storage_root(storage_key: &[u8]) -> Option<Vec<u8>> {
+pub fn child_storage_root(storage_key: &[u8]) -> Vec<u8> {
 	let mut length: u32 = 0;
 	unsafe {
 		let ptr = ext_child_storage_root.get()(storage_key.as_ptr(), storage_key.len() as u32, &mut length);
 		if length == u32::max_value() {
-			None
+			Vec::new()
```

Comment:
I would suggest keeping this to `None` or panic here. The current behavior of `ext_child_storage_root` is that it returns `None` if the storage key is invalid.

https://github.com/paritytech/substrate/blob/b51f8f2a9d420ad3b507d57f4f4a56e33b2d8ead/core/state-machine/src/ext.rs#L318

Returning `Vec::new()` here is actually really confusing for me.

Reply:
The value returned by `child_storage_root` is always a valid storage root. I see that we changed the behavior of `ext_child_storage_root` removing the valid-storage-key check. In that case, I would suggest to return the default trie root here, because `Vec::new()` is almost never a valid storage root.

Reply:
That's a really good point! I just forgot to remove this code, since `length` cannot be 0 anymore, since `ext.child_storage_root` doesn't return `Option` anymore